File: tools/image_library/download_universal_images.py
#!/usr/bin/env python3
"""
Download a large, reusable curriculum image library from Pexels.

Requirements:
- Python 3.9+
- Environment variable PEXELS_API_KEY

Usage:
  python tools/image_library/download_universal_images.py
  python tools/image_library/download_universal_images.py --total 800 --per-topic-scale 1.2
"""

from __future__ import annotations

import argparse
import csv
import json
import os
import random
import re
import time
import urllib.parse
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Tuple


PEXELS_BASE = "https://api.pexels.com/v1/search"
USER_AGENT = "LoreboundImageLibrary/1.0"


@dataclass
class Topic:
    slug: str
    label: str
    target_count: int
    queries: List[str]


def slugify(value: str) -> str:
    value = value.lower().strip()
    value = re.sub(r"[^a-z0-9]+", "-", value)
    return value.strip("-")


def load_topics(path: Path) -> Tuple[List[str], List[Topic]]:
    data = json.loads(path.read_text(encoding="utf-8"))
    styles = [str(x).strip() for x in data.get("global_styles", []) if str(x).strip()]
    topics = []
    for row in data.get("topics", []):
        topics.append(
            Topic(
                slug=str(row["slug"]).strip(),
                label=str(row["label"]).strip(),
                target_count=int(row["target_count"]),
                queries=[str(q).strip() for q in row.get("queries", []) if str(q).strip()],
            )
        )
    return styles, topics


def pexels_search(api_key: str, query: str, page: int, per_page: int = 80) -> List[dict]:
    params = urllib.parse.urlencode(
        {
            "query": query,
            "page": page,
            "per_page": per_page,
            "orientation": "landscape",
            "size": "medium",
        }
    )
    url = f"{PEXELS_BASE}?{params}"
    req = urllib.request.Request(
        url,
        headers={
            "Authorization": api_key,
            "User-Agent": USER_AGENT,
        },
    )
    with urllib.request.urlopen(req, timeout=30) as res:
        payload = json.loads(res.read().decode("utf-8"))
    return payload.get("photos", [])


def choose_image_url(photo: dict) -> str:
    src = photo.get("src", {})
    return (
        src.get("large2x")
        or src.get("large")
        or src.get("landscape")
        or src.get("medium")
        or src.get("original")
        or ""
    )


def download_binary(url: str, dest: Path) -> None:
    req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
    with urllib.request.urlopen(req, timeout=60) as res:
        content = res.read()
    dest.write_bytes(content)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--total", type=int, default=700, help="Total desired images across all topics.")
    parser.add_argument(
        "--per-topic-scale",
        type=float,
        default=1.0,
        help="Scale each topic target count before global normalization.",
    )
    parser.add_argument("--seed", type=int, default=42, help="Random seed for query shuffling.")
    parser.add_argument(
        "--out",
        type=str,
        default="assets/universal_library",
        help="Output directory for downloaded assets and manifest.",
    )
    parser.add_argument(
        "--topic-file",
        type=str,
        default="tools/image_library/topics.json",
        help="Topic configuration JSON path.",
    )
    return parser.parse_args()


def topic_budget(topics: List[Topic], total: int, per_topic_scale: float) -> Dict[str, int]:
    scaled = {t.slug: max(10, int(round(t.target_count * per_topic_scale))) for t in topics}
    sum_scaled = sum(scaled.values())
    if sum_scaled <= 0:
        return {t.slug: 0 for t in topics}
    ratio = total / sum_scaled
    budgets = {slug: max(8, int(round(v * ratio))) for slug, v in scaled.items()}
    return budgets


def main() -> int:
    args = parse_args()
    api_key = os.environ.get("PEXELS_API_KEY", "").strip()
    if not api_key:
        print("ERROR: Missing PEXELS_API_KEY in environment.")
        print("Set it, then run again.")
        return 1

    random.seed(args.seed)

    root = Path(args.out)
    root.mkdir(parents=True, exist_ok=True)
    topic_file = Path(args.topic_file)
    styles, topics = load_topics(topic_file)
    budgets = topic_budget(topics, args.total, args.per_topic_scale)

    manifest_path = root / "manifest.csv"
    seen_photo_ids = set()
    rows = []

    for topic in topics:
        want = budgets.get(topic.slug, 0)
        topic_dir = root / topic.slug
        topic_dir.mkdir(parents=True, exist_ok=True)

        query_pool = topic.queries[:]
        random.shuffle(query_pool)
        if not query_pool:
            continue

        print(f"[{topic.slug}] target={want}")
        collected = 0
        page = 1
        query_idx = 0
        attempts_without_new = 0

        while collected < want and attempts_without_new < 8:
            base_query = query_pool[query_idx % len(query_pool)]
            style_hint = styles[(query_idx + page) % len(styles)] if styles else ""
            full_query = f"{base_query} {style_hint}".strip()
            query_idx += 1

            try:
                photos = pexels_search(api_key, full_query, page=page, per_page=80)
            except Exception as exc:  # noqa: BLE001
                print(f"  WARN search failed for '{full_query}' page {page}: {exc}")
                time.sleep(1.0)
                page += 1
                continue

            if not photos:
                attempts_without_new += 1
                page += 1
                continue

            added_this_round = 0
            for photo in photos:
                pid = int(photo.get("id", 0))
                if pid <= 0 or pid in seen_photo_ids:
                    continue

                image_url = choose_image_url(photo)
                if not image_url:
                    continue

                query_key = slugify(base_query)[:36]
                fname = f"{topic.slug}__{query_key}__{pid}.jpg"
                fpath = topic_dir / fname

                try:
                    download_binary(image_url, fpath)
                except Exception as exc:  # noqa: BLE001
                    print(f"  WARN download failed id={pid}: {exc}")
                    continue

                seen_photo_ids.add(pid)
                added_this_round += 1
                collected += 1

                rows.append(
                    {
                        "file": str(fpath.as_posix()),
                        "topic": topic.slug,
                        "label": topic.label,
                        "query": base_query,
                        "tags": ";".join(
                            sorted(
                                {
                                    topic.slug,
                                    slugify(topic.label),
                                    slugify(base_query),
                                    "universal",
                                    "curriculum",
                                    "background",
                                }
                            )
                        ),
                        "photo_id": str(pid),
                        "source": "pexels",
                        "source_url": str(photo.get("url", "")),
                        "image_url": image_url,
                        "photographer": str(photo.get("photographer", "")),
                        "license_note": "Pexels License; verify usage requirements before commercial redistribution.",
                    }
                )

                if collected >= want:
                    break

            if added_this_round == 0:
                attempts_without_new += 1
            else:
                attempts_without_new = 0

            page += 1
            time.sleep(0.2)

        print(f"  downloaded={collected}")

    with manifest_path.open("w", newline="", encoding="utf-8") as fp:
        writer = csv.DictWriter(
            fp,
            fieldnames=[
                "file",
                "topic",
                "label",
                "query",
                "tags",
                "photo_id",
                "source",
                "source_url",
                "image_url",
                "photographer",
                "license_note",
            ],
        )
        writer.writeheader()
        writer.writerows(rows)

    index_by_topic: Dict[str, List[dict]] = {}
    for row in rows:
        index_by_topic.setdefault(row["topic"], []).append(
            {
                "file": row["file"],
                "tags": row["tags"].split(";"),
                "source_url": row["source_url"],
            }
        )

    summary = {
        "generated_at_unix": int(time.time()),
        "total_images": len(rows),
        "topics": {k: len(v) for k, v in index_by_topic.items()},
        "index": index_by_topic,
    }
    (root / "index.json").write_text(json.dumps(summary, indent=2), encoding="utf-8")

    print(f"\nDone. Downloaded {len(rows)} images.")
    print(f"Manifest: {manifest_path.as_posix()}")
    print(f"Index: {(root / 'index.json').as_posix()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/image_library/download_universal_images.mjs
#!/usr/bin/env node
/**
 * Free universal curriculum image pack builder.
 * Sources (no API key): Wikimedia Commons Action API + LoremFlickr.
 * Every kept image must pass: topic keyword relevance, min file size, and image mime type.
 */
import fs from "node:fs/promises";
import path from "node:path";

const USER_AGENT =
  "Mozilla/5.0 (compatible; LoreboundAssetBot/1.0; +https://localhost; curriculum pack builder)";
const WIKI_API = "https://commons.wikimedia.org/w/api.php";

function parseArgs(argv) {
  const out = {
    total: 700,
    out: "assets/universal_library",
    topicFile: "tools/image_library/topics.json",
    minBytes: 40_000,
    delayMs: 350,
  };
  for (let i = 2; i < argv.length; i += 1) {
    const key = argv[i];
    const val = argv[i + 1];
    if (key === "--total" && val) out.total = Number(val);
    if (key === "--out" && val) out.out = val;
    if (key === "--topic-file" && val) out.topicFile = val;
    if (key === "--min-bytes" && val) out.minBytes = Number(val);
    if (key === "--delay-ms" && val) out.delayMs = Number(val);
  }
  return out;
}

function slugify(value) {
  return String(value)
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function topicBudgets(topics, total) {
  const sum = topics.reduce((acc, t) => acc + Number(t.target_count || 0), 0) || 1;
  const ratio = total / sum;
  const out = {};
  for (const t of topics) out[t.slug] = Math.max(5, Math.round(t.target_count * ratio));
  return out;
}

function keywordSet(topic) {
  const raw = `${topic.slug} ${topic.label} ${(topic.queries || []).join(" ")} ${(topic.keywords || []).join(" ")}`;
  return new Set(
    raw
      .toLowerCase()
      .replace(/[^a-z0-9\s]/g, " ")
      .split(/\s+/)
      .filter((w) => w.length >= 4)
  );
}

function isRelevantTitle(title, keywords) {
  const text = String(title || "").toLowerCase();
  // Reject clearly non-image / icon / logo junk.
  if (/\.(svg|gif)$/i.test(text)) return false;
  if (/\b(icon|logo|flag|map|coat of arms|emblem|badge|seal)\b/i.test(text)) return false;
  let hits = 0;
  for (const kw of keywords) {
    if (text.includes(kw)) hits += 1;
  }
  return hits >= 1;
}

function pickExt(mime, url, title) {
  const m = String(mime || "").toLowerCase();
  if (m.includes("png")) return "png";
  if (m.includes("webp")) return "webp";
  if (m.includes("jpeg") || m.includes("jpg")) return "jpg";
  const hay = `${url || ""} ${title || ""}`.toLowerCase();
  if (hay.includes(".png")) return "png";
  if (hay.includes(".webp")) return "webp";
  return "jpg";
}

async function wikiFetchJson(params) {
  const url = new URL(WIKI_API);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, String(v));
  const res = await fetch(url, { headers: { "User-Agent": USER_AGENT } });
  if (!res.ok) throw new Error(`Wikimedia API ${res.status}`);
  return res.json();
}

async function searchWikimedia(query, limit = 40) {
  const data = await wikiFetchJson({
    action: "query",
    format: "json",
    generator: "search",
    gsrsearch: `${query} filetype:bitmap`,
    gsrnamespace: 6,
    gsrlimit: Math.min(50, limit),
    prop: "imageinfo",
    iiprop: "url|size|mime|extmetadata",
    iiurlwidth: 1600,
  });
  const pages = Object.values(data?.query?.pages || {});
  return pages
    .map((p) => {
      const ii = (p.imageinfo || [])[0] || {};
      const meta = ii.extmetadata || {};
      const title = String(p.title || "").replace(/^File:/, "");
      return {
        provider: "wikimedia",
        id: `wikimedia-${p.pageid}`,
        title,
        description: String(meta.ImageDescription?.value || "").replace(/<[^>]+>/g, " "),
        width: Number(ii.width || 0),
        height: Number(ii.height || 0),
        mimeType: String(ii.mime || ""),
        imageUrl: String(ii.thumburl || ii.url || ""),
        sourceUrl: String(ii.descriptionurl || ii.url || ""),
        author: String(meta.Artist?.value || "").replace(/<[^>]+>/g, " ").trim(),
        license: String(meta.LicenseShortName?.value || "Wikimedia Commons"),
      };
    })
    .filter((x) => x.imageUrl && /image\/(jpeg|jpg|png|webp)/i.test(x.mimeType || "image/jpeg"));
}

async function categoryWikimedia(category, limit = 40) {
  const data = await wikiFetchJson({
    action: "query",
    format: "json",
    generator: "categorymembers",
    gcmtitle: category,
    gcmtype: "file",
    gcmlimit: Math.min(50, limit),
    prop: "imageinfo",
    iiprop: "url|size|mime|extmetadata",
    iiurlwidth: 1600,
  });
  const pages = Object.values(data?.query?.pages || {});
  return pages
    .map((p) => {
      const ii = (p.imageinfo || [])[0] || {};
      const meta = ii.extmetadata || {};
      const title = String(p.title || "").replace(/^File:/, "");
      return {
        provider: "wikimedia",
        id: `wikimedia-${p.pageid}`,
        title,
        description: String(meta.ImageDescription?.value || "").replace(/<[^>]+>/g, " "),
        width: Number(ii.width || 0),
        height: Number(ii.height || 0),
        mimeType: String(ii.mime || ""),
        imageUrl: String(ii.thumburl || ii.url || ""),
        sourceUrl: String(ii.descriptionurl || ii.url || ""),
        author: String(meta.Artist?.value || "").replace(/<[^>]+>/g, " ").trim(),
        license: String(meta.LicenseShortName?.value || "Wikimedia Commons"),
      };
    })
    .filter((x) => x.imageUrl && /image\/(jpeg|jpg|png|webp)/i.test(x.mimeType || "image/jpeg"));
}

async function searchLoremFlickr(tag, count = 12) {
  const out = [];
  for (let i = 0; i < count; i += 1) {
    const lock = `${slugify(tag)}-${i + 1}`;
    const jsonUrl = `https://loremflickr.com/json/g/1400/900/${encodeURIComponent(tag)}/all?lock=${lock}`;
    try {
      const res = await fetch(jsonUrl, { headers: { "User-Agent": USER_AGENT } });
      if (!res.ok) continue;
      const data = await res.json();
      const file = String(data.file || "");
      if (!file) continue;
      out.push({
        provider: "loremflickr",
        id: `loremflickr-${lock}`,
        title: `${tag} educational scene`,
        description: `${tag} universal background`,
        width: Number(data.width || 1400),
        height: Number(data.height || 900),
        mimeType: "image/jpeg",
        imageUrl: file,
        sourceUrl: file,
        author: "LoremFlickr / Flickr CC sources",
        license: "Check Flickr license of source photo",
      });
    } catch {
      // ignore single failures
    }
  }
  return out;
}

async function downloadBinary(url, dest) {
  const res = await fetch(url, {
    headers: { "User-Agent": USER_AGENT, Accept: "image/*,*/*;q=0.8" },
    redirect: "follow",
  });
  if (!res.ok) throw new Error(`Download ${res.status}`);
  const ct = (res.headers.get("content-type") || "").toLowerCase();
  if (ct && !ct.includes("image/") && !ct.includes("octet-stream")) {
    throw new Error(`Not an image content-type: ${ct}`);
  }
  const bytes = Buffer.from(await res.arrayBuffer());
  await fs.writeFile(dest, bytes);
  return { bytes: bytes.length, contentType: ct, finalUrl: res.url };
}

async function collectCandidates(topic, delayMs) {
  const candidates = [];
  const categories = topic.wikimedia_categories || [];
  const queries = topic.queries || [topic.slug];

  for (const category of categories.slice(0, 3)) {
    try {
      const rows = await categoryWikimedia(category, 40);
      candidates.push(...rows);
      await sleep(delayMs);
    } catch (err) {
      console.warn(`  WARN category ${category}: ${err.message}`);
      await sleep(delayMs * 2);
    }
  }

  for (const q of queries.slice(0, 3)) {
    try {
      const rows = await searchWikimedia(q, 30);
      candidates.push(...rows);
      await sleep(delayMs);
    } catch (err) {
      console.warn(`  WARN search ${q}: ${err.message}`);
      await sleep(delayMs * 2);
    }
  }

  // Free no-key keyword images as fill for universal scene coverage.
  for (const q of queries.slice(0, 2)) {
    const tag = q.split(/\s+/).slice(0, 2).join(",");
    try {
      const rows = await searchLoremFlickr(tag, 8);
      candidates.push(...rows);
      await sleep(delayMs);
    } catch (err) {
      console.warn(`  WARN loremflickr ${tag}: ${err.message}`);
    }
  }

  return candidates;
}

async function main() {
  const args = parseArgs(process.argv);
  const config = JSON.parse(await fs.readFile(args.topicFile, "utf8"));
  const topics = config.topics || [];
  const budgets = topicBudgets(topics, args.total);

  await fs.mkdir(args.out, { recursive: true });
  const seen = new Set();
  const rows = [];

  for (const topic of topics) {
    const want = budgets[topic.slug] || 0;
    const topicDir = path.join(args.out, topic.slug);
    await fs.mkdir(topicDir, { recursive: true });
    const keywords = keywordSet(topic);
    console.log(`[${topic.slug}] target=${want}`);

    const candidates = await collectCandidates(topic, args.delayMs);
    let collected = 0;

    for (const item of candidates) {
      if (collected >= want) break;
      const pid = String(item.id || "");
      if (!pid || seen.has(pid)) continue;

      const titleBlob = `${item.title} ${item.description}`;
      // LoremFlickr tags are intentional by query, Wikimedia needs title relevance.
      if (item.provider === "wikimedia" && !isRelevantTitle(titleBlob, keywords)) continue;
      if (!item.imageUrl) continue;
      if (item.width && item.width < 700) continue;
      if (item.height && item.height < 400) continue;

      const ext = pickExt(item.mimeType, item.imageUrl, item.title);
      const queryKey = slugify((topic.queries && topic.queries[0]) || topic.slug).slice(0, 36);
      const idTail = String(pid).replace(/^wikimedia-/, "w-").replace(/^loremflickr-/, "lf-");
      const filename = `${topic.slug}__${queryKey}__${slugify(idTail)}.${ext}`;
      const filePath = path.join(topicDir, filename);

      try {
        const meta = await downloadBinary(item.imageUrl, filePath);
        if (meta.bytes < args.minBytes) {
          await fs.unlink(filePath).catch(() => {});
          continue;
        }
        // Quick magic-byte / content-type double check: tiny non-image leftovers.
        if (meta.contentType.includes("svg") || meta.contentType.includes("html")) {
          await fs.unlink(filePath).catch(() => {});
          continue;
        }
        seen.add(pid);
        collected += 1;
        rows.push({
          file: filePath.replaceAll("\\", "/"),
          topic: topic.slug,
          label: topic.label,
          query: topic.queries?.[0] || topic.slug,
          tags: [topic.slug, slugify(topic.label), "universal", "curriculum", "background"].sort().join(";"),
          photo_id: pid,
          source: item.provider,
          source_url: item.sourceUrl,
          image_url: item.imageUrl,
          photographer: item.author,
          license_note: item.license,
          bytes: String(meta.bytes),
        });
        process.stdout.write(`  + ${filename} (${Math.round(meta.bytes / 1024)}KB)\n`);
      } catch (err) {
        console.warn(`  WARN download failed ${pid}: ${err.message}`);
      }
      await sleep(args.delayMs);
    }

    console.log(`  downloaded=${collected}`);
  }

  const header = [
    "file",
    "topic",
    "label",
    "query",
    "tags",
    "photo_id",
    "source",
    "source_url",
    "image_url",
    "photographer",
    "license_note",
    "bytes",
  ];
  const csv = [header.join(",")]
    .concat(
      rows.map((row) =>
        header
          .map((k) => `"${String(row[k] || "").replaceAll('"', '""')}"`)
          .join(",")
      )
    )
    .join("\n");
  await fs.writeFile(path.join(args.out, "manifest.csv"), csv, "utf8");

  const index = {};
  for (const row of rows) {
    if (!index[row.topic]) index[row.topic] = [];
    index[row.topic].push({
      file: row.file,
      tags: row.tags.split(";"),
      source: row.source,
      source_url: row.source_url,
      bytes: Number(row.bytes || 0),
    });
  }
  await fs.writeFile(
    path.join(args.out, "index.json"),
    JSON.stringify(
      {
        generated_at_unix: Math.floor(Date.now() / 1000),
        total_images: rows.length,
        topics: Object.fromEntries(Object.entries(index).map(([k, v]) => [k, v.length])),
        index,
      },
      null,
      2
    ),
    "utf8"
  );

  console.log(`\nDone. Downloaded ${rows.length} verified images.`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
